src/ralph/dns/dnsaas.py

```diff
@@ -1,39 +1,89 @@
 # -*- coding: utf-8 -*-
 import json
 import logging
-from urllib.parse import urlencode, urljoin
+from urllib.parse import urlencode, urljoin, parse_qs, urlsplit
+from datetime import datetime, timedelta
 
 import requests
+from dj.choices import Choices
 from django.conf import settings
 from django.utils.translation import ugettext_lazy as _
-
-from ralph.dns.forms import RecordType
-from ralph.helpers import cache
+from oauthlib.oauth2 import BackendApplicationClient
+from oauthlib.oauth2.rfc6749.errors import CustomOAuth2Error
+from requests_oauthlib import OAuth2Session
+from typing import Optional, List, Tuple
 
 logger = logging.getLogger(__name__)
 
+QueryParams = List[Tuple[str, str]]
+
+
+class RecordType(Choices):
+    _ = Choices.Choice
+
+    a = _('A')
+    txt = _('TXT')
+    cname = _('CNAME')
+
 
 class DNSaaS:
 
-    def __init__(self, headers=None):
+    def __init__(self, headers: dict = None):
         self.session = requests.Session()
         _headers = {
-            'Authorization': 'Token {}'.format(settings.DNSAAS_TOKEN),
             'Content-Type': 'application/json',
-            'User-agent': 'Ralph/DNSaaS/Client'
+            'User-agent': 'Ralph/DNSaaS/Client',
+            'Authorization': 'Bearer {}'.format(self._get_oauth_token())
         }
         if headers is not None:
             _headers.update(headers)
         self.session.headers.update(_headers)
 
-    def build_url(self, resource_name, id=None, version='v2', get_params=None):
+    def _get_oauth_token(self):
+        client_id = settings.OAUTH_CLIENT_ID
+        secret = settings.OAUTH_SECRET
+        token_url = settings.OAUTH_TOKEN_URL
+        client = BackendApplicationClient(client_id=client_id)
+        oauth = OAuth2Session(client=client)
+        try:
+            token = oauth.fetch_token(
+                token_url=token_url, client_id=client_id, client_secret=secret
+            )
+        except CustomOAuth2Error as e:
+            logger.error(str(e))
+
+        expire_in = token.get('expires_in')
+        self.token_expiration = datetime.now() + timedelta(0, expire_in - 60)
+        return token.get('access_token')
+
+    def _update_oauth_token(self):
+        token = self._get_oauth_token()
+        self.session.headers['Authorization'] = 'Bearer {}'.format(token)
+
+    def _verify_oauth_token_validity(self):
+        if datetime.now() >= self.token_expiration:
+            self._update_oauth_token()
+
+    def renew_token_when_unauthorized(self, func):
+        self._verify_oauth_token_validity()
```

Comment:
To niestety nie jest poprawna konstrukcja, tutaj ta weryfikacja oauth tokena odbędzie się tylko raz. Na etapie wczytania pythonowego pliku podczas interpretowania dekoratora @renew_token_when_unauthorized. Każde kolejne wywołanie dekorowanej funkcji nie uruchomi tej walidacji